visions/webassembly.md

```diff
@@ -0,0 +1,94 @@
+address more than 4 GiB of memory. Support for 64-bit pointers was added to the WebAssembly spec and widely used
+runtimes later as an extension. Wasm binaries utilizing certain pointer width are referred as supporting `wasm32`
+and `wasm64` "architectures" respectively.
+
+### WebAssembly System Interface (WASI)
+
+WebAssembly instruction set on its own doesn't "support" file I/O or networking, in the same way that ARM64 or x86_64
+don't "support" those directly either. Actual implementation of I/O for a hardware CPU is provided by the operating
+system, and for a Wasm module it's provided by a runtime that executes it.
+
+A standardized set of APIs implemented by a Wasm runtime for interaction with the host operating system is called
+[WebAssembly System Interface](https://wasi.dev). A layer on top of WASI that Swift apps compiled to Wasm can already
+use thanks to C interop is [WASI libc](https://github.com/WebAssembly/wasi-libc). In fact, the current implementation of
+Swift stdlib and runtime for `wasm32-unknown-wasi` triple is based on this C library.
+
+### The WebAssembly Component Model
+
+Initial version of WASI (referred to as "Preview 1" or as `wasi_snapshot_preview1` by its module name) was
+inspired by C ABI and POSIX, and WASI libc itself is a fork of [Musl libc](http://musl.libc.org) originally developed for
+Linux. This proved to be limiting with continued development of WASI, especially as it does not necessarily have to
+be constrained by C ABI and POSIX. A more powerful runtime implementation can abstract these away.
+
+At the same time, W3C WebAssembly Working Group was considering multiple proposals for improving the WebAssembly [type
+system](https://github.com/webassembly/interface-types) and
+[module linking](https://github.com/webassembly/module-linking). These were later subsumed into a combined
+[Component Model](https://component-model.bytecodealliance.org) proposal thanks to the ongoing work on
+[WASI Preview 2](https://github.com/WebAssembly/WASI/blob/main/preview2/README.md), which served as playground for
+the new design.
+
+The Component Model defines these core concepts:
+
+- A *component* is a composable container for one or more WebAssembly modules that have a predefined interface;
+- *WebAssembly Interface Types (WIT) language* allows defining contracts between components;
+- *Canonical ABI* is an ABI for types defined by WIT and used by component interfaces in the Component Model.
+
+WIT is a high-level language with
+[an advanced type system](https://component-model.bytecodealliance.org/design/wit.html#built-in-types). It can be
+particularly interesting for Swift, as it allows significantly more Swift APIs to be exposed directly in interfaces of
+Wasm components compiled from Swift.
+
+Preliminary support for WIT has been implemented in
+[the `wit-tool` subcommand](https://github.com/swiftwasm/WasmKit/blob/0.0.3/Sources/WITTool/WITTool.swift) of WasmKit
+CLI.
+
+## Goals
+
+TBD
```

Comment:
Potential goals:

- Make it easier to bring existing ecosystem assets to WebAssembly use cases
- Make the language to be capable to 
- Provide better tooling support for cross-compilation (because Wasm targets always require cross-compilation unlike other targets)